--- b/dlls/ntdll/signal_arm64.c	2026-04-02 23:10:32.548987769 +0800
+++ download_hangover_11_4/dlls/ntdll/signal_arm64.c	2026-04-02 22:55:13.893770392 +0800
@@ -28,7 +28,6 @@
 #include <setjmp.h>
 
 #include "ntstatus.h"
-#define WIN32_NO_STATUS
 #include "windef.h"
 #include "winternl.h"
 #include "ddk/wdm.h"
@@ -182,6 +181,19 @@
                    "ret" )
 
 
+
+
+/*******************************************************************
+ *         nested_exception_handler
+ */
+EXCEPTION_DISPOSITION WINAPI nested_exception_handler( EXCEPTION_RECORD *rec, void *frame,
+                                                       CONTEXT *context, void *dispatch )
+{
+    if (rec->ExceptionFlags & (EXCEPTION_UNWINDING | EXCEPTION_EXIT_UNWIND)) return ExceptionContinueSearch;
+    return ExceptionNestedException;
+}
+
+
 /***********************************************************************
  *		call_seh_handler
  */
@@ -601,6 +613,16 @@
 }
 
 
+static ULONGLONG cpu_features_bitmap[2];
+static RTL_RUN_ONCE init_once = RTL_RUN_ONCE_INIT;
+
+static DWORD WINAPI init_cpu_features( RTL_RUN_ONCE *once, void *param, void **context )
+{
+    return !NtQuerySystemInformation( SystemProcessorFeaturesBitMapInformation,
+                                      cpu_features_bitmap, sizeof(cpu_features_bitmap), NULL );
+}
+
+
 /***********************************************************************
  *           RtlIsProcessorFeaturePresent [NTDLL.@]
  */
@@ -637,10 +659,25 @@
         (1ull << PF_ARM_SVE_SM4_INSTRUCTIONS_AVAILABLE) |
         (1ull << PF_ARM_SVE_I8MM_INSTRUCTIONS_AVAILABLE) |
         (1ull << PF_ARM_SVE_F32MM_INSTRUCTIONS_AVAILABLE) |
-        (1ull << PF_ARM_SVE_F64MM_INSTRUCTIONS_AVAILABLE);
+        (1ull << PF_ARM_SVE_F64MM_INSTRUCTIONS_AVAILABLE) |
+        (1ull << PF_ARM_LSE2_AVAILABLE);
 
-    return (feature < PROCESSOR_FEATURE_MAX && (arm64_features & (1ull << feature)) &&
-            user_shared_data->ProcessorFeatures[feature]);
+    if (feature < PROCESSOR_FEATURE_MAX)
+        return (arm64_features & (1ull << feature)) && user_shared_data->ProcessorFeatures[feature];
+
+    feature -= PROCESSOR_FEATURE_MAX;
+    if (feature >= 8 * sizeof(cpu_features_bitmap)) return FALSE;
+
+    RtlRunOnceExecuteOnce( &init_once, init_cpu_features, NULL, NULL );
+    return !!(cpu_features_bitmap[feature / 64] & (1ull << (feature % 64)));
+}
+
+static void suspend_remote_breakin( HANDLE thread )
+{
+    ULONG count;
+    NTSTATUS status = pWow64SuspendLocalThread( thread, &count );
+    if (status >= 0) status = count;
+    NtTerminateThread( GetCurrentThread(),  status );
 }
 
 /***********************************************************************
@@ -648,14 +685,56 @@
  */
 NTSTATUS WINAPI RtlWow64SuspendThread( HANDLE thread, ULONG *count )
 {
+    HANDLE thread_dup;
     THREAD_BASIC_INFORMATION tbi;
+    NTSTATUS status = NtDuplicateObject( NtCurrentProcess(), thread, NtCurrentProcess(), &thread_dup,
+                                         THREAD_QUERY_INFORMATION | THREAD_SUSPEND_RESUME, 0, 0 );
+    if (status) return status;
+    status = NtQueryInformationThread( thread_dup, ThreadBasicInformation, &tbi, sizeof(tbi), NULL);
+    NtClose( thread_dup );
+    if (status) return status;
 
-    NTSTATUS ret = NtQueryInformationThread( thread, ThreadBasicInformation, &tbi, sizeof(tbi), NULL);
-    if (ret) return ret;
+    if (tbi.ClientId.UniqueProcess != NtCurrentTeb()->ClientId.UniqueProcess)
+    {
+        HANDLE process;
+        HANDLE suspender_thread;
+        HANDLE remote_suspendee_thread;
+        OBJECT_ATTRIBUTES attr = { .Length = sizeof(attr) };
+
+        status = NtOpenProcess( &process, PROCESS_CREATE_THREAD | PROCESS_DUP_HANDLE, &attr, &tbi.ClientId );
+        if (status) return status;
+
+        status = NtDuplicateObject( NtCurrentProcess(), thread, process, &remote_suspendee_thread, 0, 0,
+                                    DUPLICATE_SAME_ACCESS );
+        if (status) goto err_close_proc;
+
+        status = NtCreateThreadEx( &suspender_thread, SYNCHRONIZE | THREAD_QUERY_INFORMATION, NULL, process,
+                                   suspend_remote_breakin, remote_suspendee_thread,
+                                   THREAD_CREATE_FLAGS_SKIP_THREAD_ATTACH | THREAD_CREATE_FLAGS_SKIP_LOADER_INIT |
+                                   THREAD_CREATE_FLAGS_HIDE_FROM_DEBUGGER | THREAD_CREATE_FLAGS_BYPASS_PROCESS_FREEZE,
+                                   0, 0, 0, NULL );
+        if (status) goto err_close_remote_hnd;
+
+        NtWaitForSingleObject( suspender_thread, FALSE, NULL );
+        status = NtQueryInformationThread( suspender_thread, ThreadBasicInformation, &tbi, sizeof(tbi), NULL );
+        if (!status)
+        {
+            if (tbi.ExitStatus < 0)
+            {
+                status = tbi.ExitStatus;
+            }
+            else if (count)
+            {
+                *count = (ULONG)tbi.ExitStatus;
+            }
+        }
 
-    if (tbi.ClientId.UniqueProcess != NtCurrentTeb()->ClientId.UniqueProcess) {
-        FIXME( "Non-local process thread suspend\n" );
-        return STATUS_SUCCESS;
+        NtClose( suspender_thread );
+err_close_remote_hnd:
+        NtDuplicateObject( process, remote_suspendee_thread, NULL, NULL, 0, 0, DUPLICATE_CLOSE_SOURCE );
+err_close_proc:
+        NtClose( process );
+        return status;
     }
 
     return pWow64SuspendLocalThread( thread, count );
