diff -urN wine_proton10/dlls/dinput8/Makefile.in wine-10.0/dlls/dinput8/Makefile.in
--- wine_proton10/dlls/dinput8/Makefile.in	2025-11-05 17:27:18.578954007 +0800
+++ wine-10.0/dlls/dinput8/Makefile.in	2025-01-22 00:42:16.000000000 +0800
@@ -2,7 +2,6 @@
 IMPORTLIB = dinput8
 IMPORTS   = dinput8 dxguid uuid comctl32 ole32 user32 advapi32 hid setupapi
 EXTRADEFS = -DDIRECTINPUT_VERSION=0x0800
-EXTRADLLFLAGS = -Wb,--prefer-native
 PARENTSRC = ../dinput
 
 SOURCES = \
