--- wine_proton10/programs/services/services.c	2025-11-05 17:27:22.020948047 +0800
+++ wine-10.0/programs/services/services.c	2025-01-22 00:42:16.000000000 +0800
@@ -172,7 +172,7 @@
     return ERROR_SUCCESS;
 }
 
-static DWORD reg_set_string_value(HKEY hKey, LPCWSTR value_name, LPCWSTR string, BOOL expand)
+static DWORD reg_set_string_value(HKEY hKey, LPCWSTR value_name, LPCWSTR string)
 {
     if (!string)
     {
@@ -184,8 +184,7 @@
         return ERROR_SUCCESS;
     }
 
-    return RegSetValueExW(hKey, value_name, 0, expand ? REG_EXPAND_SZ : REG_SZ,
-            (const BYTE*)string, sizeof(WCHAR)*(lstrlenW(string) + 1));
+    return RegSetValueExW(hKey, value_name, 0, REG_SZ, (const BYTE*)string, sizeof(WCHAR)*(lstrlenW(string) + 1));
 }
 
 static DWORD reg_set_multisz_value(HKEY hKey, LPCWSTR value_name, LPCWSTR string)
@@ -221,11 +220,11 @@
     if (err != ERROR_SUCCESS)
         goto cleanup;
 
-    if ((err = reg_set_string_value(hKey, L"DisplayName", entry->config.lpDisplayName, FALSE))) goto cleanup;
-    if ((err = reg_set_string_value(hKey, L"ImagePath", entry->config.lpBinaryPathName, TRUE))) goto cleanup;
-    if ((err = reg_set_string_value(hKey, L"Group", entry->config.lpLoadOrderGroup, FALSE))) goto cleanup;
-    if ((err = reg_set_string_value(hKey, L"ObjectName", entry->config.lpServiceStartName, FALSE))) goto cleanup;
-    if ((err = reg_set_string_value(hKey, L"Description", entry->description, FALSE))) goto cleanup;
+    if ((err = reg_set_string_value(hKey, L"DisplayName", entry->config.lpDisplayName))) goto cleanup;
+    if ((err = reg_set_string_value(hKey, L"ImagePath", entry->config.lpBinaryPathName))) goto cleanup;
+    if ((err = reg_set_string_value(hKey, L"Group", entry->config.lpLoadOrderGroup))) goto cleanup;
+    if ((err = reg_set_string_value(hKey, L"ObjectName", entry->config.lpServiceStartName))) goto cleanup;
+    if ((err = reg_set_string_value(hKey, L"Description", entry->description))) goto cleanup;
     if ((err = reg_set_multisz_value(hKey, L"DependOnService", entry->dependOnServices))) goto cleanup;
     if ((err = reg_set_multisz_value(hKey, L"DependOnGroup", entry->dependOnGroups))) goto cleanup;
     if ((err = reg_set_dword_value(hKey, L"Start", entry->config.dwStartType))) goto cleanup;
@@ -1157,12 +1156,6 @@
     BOOL shared_process;
     DWORD err;
 
-    if (service->name && (!wcscmp(service->name, L"edgeupdate") || !wcscmp(service->name, L"edgeupdatem")))
-    {
-        ERR( "HACK: not starting %s.\n", wine_dbgstr_w(service->name));
-        return ERROR_SUCCESS;
-    }
-
     err = service_start_process(service, &process, &shared_process);
     if (err == ERROR_SUCCESS)
     {
