From 14b3e5684c73d041375fd83143284162b59f1ea4 Mon Sep 17 00:00:00 2001
From: BrunoSX <brunodeveloper85@gmail.com>
Date: Thu, 25 Sep 2025 16:46:45 -0300
Subject: [PATCH 01/10] win32u: Do not duplicate explorer

---
 dlls/win32u/winstation.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/dlls/win32u/winstation.c b/dlls/win32u/winstation.c
index eaa5f40..abd14b3 100755
--- a/dlls/win32u/winstation.c
+++ b/dlls/win32u/winstation.c
@@ -45,6 +45,8 @@ WINE_DECLARE_DEBUG_CHANNEL(win);
 
 #define DESKTOP_ALL_ACCESS 0x01ff
 
+static const WCHAR winsta0[] = {'W','i','n','S','t','a','0',0};
+
 struct shared_input_cache
 {
     const shared_object_t *object;
@@ -875,6 +877,10 @@ HWND get_desktop_window(void)
             NtResumeThread( thread, NULL );
             TRACE_(win)( "started explorer\n" );
             NtUserWaitForInputIdle( process, 10000, FALSE );
+
+            if (!wcsncmp( winsta0, desktop, wcslen(winsta0) ))
+                NtTerminateProcess( process, 0 );
+
             NtClose( thread );
             NtClose( process );
         }
@@ -976,8 +982,6 @@ void winstation_init(void)
     OBJECT_ATTRIBUTES attr;
     UNICODE_STRING str;
 
-    static const WCHAR winsta0[] = {'W','i','n','S','t','a','0',0};
-
     if (params->Desktop.Length)
     {
         buffer = malloc( params->Desktop.Length + sizeof(WCHAR) );
-- 
2.53.0

