From 08c8029e43cbf303f2ac9c1b0ba861bb43938ebc Mon Sep 17 00:00:00 2001
From: BrunoSX <brunodeveloper85@gmail.com>
Date: Sat, 1 Nov 2025 16:28:10 -0300
Subject: [PATCH 08/10] explorer: Fix nogui option

---
 programs/explorer/desktop.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/programs/explorer/desktop.c b/programs/explorer/desktop.c
index e3c92f2..6ef7cf5 100755
--- a/programs/explorer/desktop.c
+++ b/programs/explorer/desktop.c
@@ -1301,7 +1301,7 @@ void manage_desktop( WCHAR *arg )
         initialize_display_settings( width, height );
         initialize_appbar();
 
-        initialize_systray( using_root || nogui, enable_shell, show_systray, no_tray_items );
+        initialize_systray( using_root, enable_shell, show_systray, no_tray_items );
         if (!using_root && enable_launchers) initialize_launchers( hwnd );
 
         if ((shell32 = LoadLibraryW( L"shell32.dll" )) &&
-- 
2.53.0

