From 7005b80d4a732fcf8982a03265e205ba29781454 Mon Sep 17 00:00:00 2001
From: BrunoSX <brunodeveloper85@gmail.com>
Date: Tue, 23 Sep 2025 15:05:51 -0300
Subject: [PATCH 22/40] explorer: Add nogui option

---
 programs/explorer/desktop.c | 27 +++++++++++++++++++--------
 1 file changed, 19 insertions(+), 8 deletions(-)

diff --git a/programs/explorer/desktop.c b/programs/explorer/desktop.c
index 693e2ff..e3c92f2 100755
--- a/programs/explorer/desktop.c
+++ b/programs/explorer/desktop.c
@@ -43,6 +43,7 @@ WINE_DEFAULT_DEBUG_CHANNEL(explorer);
 static const WCHAR default_driver[] = L"mac,x11,wayland";
 
 static BOOL using_root = TRUE;
+static BOOL nogui = FALSE;
 
 struct launcher
 {
@@ -816,7 +817,7 @@ static LRESULT WINAPI desktop_wnd_proc( HWND hwnd, UINT message, WPARAM wp, LPAR
         return HTCLIENT;
 
     case WM_ERASEBKGND:
-        if (!using_root) PaintDesktop( (HDC)wp );
+        if (!using_root && !nogui) PaintDesktop( (HDC)wp );
         return TRUE;
 
     case WM_SETTINGCHANGE:
@@ -829,7 +830,7 @@ static LRESULT WINAPI desktop_wnd_proc( HWND hwnd, UINT message, WPARAM wp, LPAR
         return 0;
 
     case WM_LBUTTONDBLCLK:
-        if (!using_root)
+        if (!using_root && !nogui)
         {
             const struct launcher *launcher = launcher_from_point( (short)LOWORD(lp), (short)HIWORD(lp) );
             if (launcher) ShellExecuteW( NULL, L"open", launcher->path, NULL, NULL, 0 );
@@ -840,7 +841,7 @@ static LRESULT WINAPI desktop_wnd_proc( HWND hwnd, UINT message, WPARAM wp, LPAR
         {
             PAINTSTRUCT ps;
             BeginPaint( hwnd, &ps );
-            if (!using_root)
+            if (!using_root && !nogui)
             {
                 PaintDesktop( ps.hdc );
                 draw_launchers( ps.hdc, ps.rcPaint );
@@ -920,6 +921,9 @@ static BOOL get_default_enable_shell( const WCHAR *name )
     BOOL result;
     DWORD size = sizeof(result);
 
+    /* For the magic desktop name "shell" return TRUE */
+    if (!lstrcmpiW( name, L"shell" )) return TRUE;    
+
     /* @@ Wine registry key: HKCU\Software\Wine\Explorer\Desktops */
     if (!RegOpenKeyW( HKEY_CURRENT_USER, L"Software\\Wine\\Explorer\\Desktops", &hkey ))
     {
@@ -927,7 +931,8 @@ static BOOL get_default_enable_shell( const WCHAR *name )
             found = TRUE;
         RegCloseKey( hkey );
     }
-    /* Default off, except for the magic desktop name "shell" */
+
+    /* Default off */
     if (!found) result = (lstrcmpiW( name, L"shell" ) == 0);
     return result;
 }
@@ -1239,9 +1244,15 @@ void manage_desktop( WCHAR *arg )
         if (!get_default_desktop_size( name, &width, &height )) width = height = 0;
     }
 
-    enable_shell = name ? get_default_enable_shell( name ) : FALSE;
-    enable_launchers = get_default_enable_launchers();
-    show_systray = get_default_show_systray( name );
+    if (name)
+    {
+        nogui = !wcsicmp( name, L"nogui" );
+        enable_shell = nogui || get_default_enable_shell( name );
+    }
+    else enable_shell = FALSE;
+
+    enable_launchers = !nogui && get_default_enable_launchers();
+    show_systray = !nogui && get_default_show_systray( name );
     no_tray_items = get_no_tray_items_display();
 
     UuidCreate( &guid );
@@ -1290,7 +1301,7 @@ void manage_desktop( WCHAR *arg )
         initialize_display_settings( width, height );
         initialize_appbar();
 
-        initialize_systray( using_root, enable_shell, show_systray, no_tray_items );
+        initialize_systray( using_root || nogui, enable_shell, show_systray, no_tray_items );
         if (!using_root && enable_launchers) initialize_launchers( hwnd );
 
         if ((shell32 = LoadLibraryW( L"shell32.dll" )) &&
-- 
2.53.0

