diff --color -urN wine-10.15/loader/preloader.c b/loader/preloader.c
--- wine-10.15/loader/preloader.c	2026-03-11 20:17:01.990985602 +0800
+++ b/loader/preloader.c	2026-03-12 19:11:52.152226206 +0800
@@ -1449,7 +1449,7 @@
         {
             /* don't warn for low 64k */
             if (preload_info[i].addr >= (void *)0x10000
-#ifdef __aarch64__
+#if 1
                 && preload_info[i].addr < (void *)0x7fffffffff /* ARM64 address space might end here*/
 #endif
             )
