diff --color -urN b/dlls/xinput1_1/Makefile.in b_fixed/dlls/xinput1_1/Makefile.in
--- b/dlls/xinput1_1/Makefile.in	2026-04-23 22:49:14.075460628 +0800
+++ b_fixed/dlls/xinput1_1/Makefile.in	2026-04-23 23:01:33.969952748 +0800
@@ -1,6 +1,6 @@
 EXTRADEFS = -DXINPUT_VER=1
 MODULE    = xinput1_1.dll
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
 PARENTSRC = ../xinput1_3
 
 SOURCES = \
