diff -urN proton10_wine/dlls/midimap/Makefile.in b/dlls/midimap/Makefile.in
--- proton10_wine/dlls/midimap/Makefile.in	2025-10-26 19:03:17.463768487 +0800
+++ b/dlls/midimap/Makefile.in	2025-10-26 21:03:26.486285582 +0800
@@ -1,5 +1,5 @@
 MODULE    = midimap.dll
-IMPORTS   = winmm user32 advapi32
+IMPORTS   = winmm user32 advapi32 ws2_32
 
 SOURCES = \
 	midimap.c
diff -urN proton10_wine/dlls/midimap/midimap.spec b/dlls/midimap/midimap.spec
--- proton10_wine/dlls/midimap/midimap.spec	2025-10-26 19:03:17.463768487 +0800
+++ b/dlls/midimap/midimap.spec	2025-10-26 21:20:27.027518448 +0800
@@ -1,3 +1,4 @@
 @ stdcall -private DriverProc(long long long long long) MIDIMAP_DriverProc
 @ stdcall -private modMessage(long long long long long) MIDIMAP_modMessage
+@ stdcall -private midMessage(long long long long long) MIDIMAP_midMessage
 @ stub modmCallback
