diff -urN proton10_wine/dlls/crypt32/unixlib.c b/dlls/crypt32/unixlib.c
--- proton10_wine/dlls/crypt32/unixlib.c	2025-10-26 19:03:16.477770194 +0800
+++ b/dlls/crypt32/unixlib.c	2025-10-26 20:18:58.647905141 +0800
@@ -621,6 +621,7 @@
  "/usr/local/share/certs/",
  "/etc/sfw/openssl/certs",
  "/etc/security/cacerts",  /* Android */
+ "/data/data/com.winlator/files/rootfs/etc/ca-certificates/cacert.pem",
 };
 
 static void load_root_certs(void)
diff -urN proton10_wine/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
--- proton10_wine/dlls/ntdll/unix/server.c	2025-10-26 19:03:17.868767786 +0800
+++ b/dlls/ntdll/unix/server.c	2025-10-26 20:14:39.162354456 +0800
@@ -1266,7 +1266,7 @@
 #ifdef __ANDROID__  /* there's no /tmp dir on Android */
     asprintf( &dir, "%s/.wineserver/server-%llx-%llx", config_dir, (unsigned long long)dev, (unsigned long long)ino );
 #else
-    asprintf( &dir, "/tmp/.wine-%u/server-%llx-%llx", getuid(), (unsigned long long)dev, (unsigned long long)ino );
+    asprintf( &dir, "/data/data/com.winlator/files/rootfs/tmp/.wine-%u/server-%llx-%llx", getuid(), (unsigned long long)dev, (unsigned long long)ino );
 #endif
     return dir;
 }
@@ -1309,7 +1309,7 @@
     {
         mkdir( "drive_c", 0777 );
         symlink( "../drive_c", "dosdevices/c:" );
-        symlink( "/", "dosdevices/z:" );
+        symlink( "/data/data/com.winlator/files/rootfs", "dosdevices/z:" );
     }
     else if (errno != EEXIST) fatal_perror( "cannot create %s/dosdevices", config_dir );
 
Binary files proton10_wine/.git/index and b/.git/index differ
diff -urN proton10_wine/programs/winemenubuilder/winemenubuilder.c b/programs/winemenubuilder/winemenubuilder.c
--- proton10_wine/programs/winemenubuilder/winemenubuilder.c	2025-10-26 19:03:20.202763744 +0800
+++ b/programs/winemenubuilder/winemenubuilder.c	2025-10-26 20:31:47.995572953 +0800
@@ -1621,7 +1621,7 @@
     if (_wgetenv( L"XDG_DATA_DIRS" ))
         dirs = xwcsdup( _wgetenv( L"XDG_DATA_DIRS" ));
     else
-        dirs = xwcsdup( L"/usr/local/share/:/usr/share/" );
+        dirs = xwcsdup( L"/data/data/com.winlator/files/rootfs/usr/share:/data/data/com.winlator/files/rootfs/usr/local/share" );
 
     ret = add_mimes(xdg_data_dir, mime_types);
     if (ret)
diff -urN proton10_wine/server/request.c b/server/request.c
--- proton10_wine/server/request.c	2025-10-26 19:03:20.259763645 +0800
+++ b/server/request.c	2025-10-26 20:12:42.546556385 +0800
@@ -645,7 +645,7 @@
     if (asprintf( &base_dir, "%s/.wineserver", config_dir ) == -1)
         fatal_error( "out of memory\n" );
 #else
-    if (asprintf( &base_dir, "/tmp/.wine-%u", getuid() ) == -1)
+    if (asprintf( &base_dir, "/data/data/com.winlator/files/rootfs/tmp/.wine-%u", getuid() ) == -1)
         fatal_error( "out of memory\n" );
 #endif
     create_dir( base_dir, &st2 );
