diff -urN proton10_wine/dlls/winmm/lolvldrv.c b/dlls/winmm/lolvldrv.c
--- proton10_wine/dlls/winmm/lolvldrv.c	2025-10-26 19:03:18.815766146 +0800
+++ b/dlls/winmm/lolvldrv.c	2025-10-27 00:01:00.256837783 +0800
@@ -361,6 +361,10 @@
     /* got some drivers */
     if (lpDrv->bIsMapper) {
         llTypes[type].nMapper = MMDrvsHi;
+        if (type == MMDRV_MIDIIN || type == MMDRV_MIDIOUT) {
+            llTypes[type].wMaxId = 1;
+            part->nIDMax = 1;
+        }
     } else {
 	if (count == 0)
 	    return FALSE;
