From 20c833d54ffec14624b48e53d22d69ebbf3a94a7 Mon Sep 17 00:00:00 2001
From: BrunoSX <brunodeveloper85@gmail.com>
Date: Tue, 17 Jun 2025 14:32:03 -0300
Subject: [PATCH] sechost: Add WINE_DO_NOT_OPEN_SC_MANAGER environment variable

---
 dlls/sechost/service.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/dlls/sechost/service.c b/dlls/sechost/service.c
index 240e048a..b4e921b5 100755
--- a/dlls/sechost/service.c
+++ b/dlls/sechost/service.c
@@ -271,9 +271,13 @@ SC_HANDLE WINAPI DECLSPEC_HOTPATCH OpenSCManagerW( const WCHAR *machine, const W
 {
     SC_RPC_HANDLE handle = NULL;
     DWORD err;
+    char do_not_open[2] = {0};
 
     TRACE( "%s %s %#lx\n", debugstr_w(machine), debugstr_w(database), access );
 
+    if (GetEnvironmentVariableA( "WINE_DO_NOT_OPEN_SC_MANAGER", do_not_open, 2 ) && do_not_open[0] == '1')
+        return NULL;
+
     __TRY
     {
         err = svcctl_OpenSCManagerW( machine, database, access, &handle );
