diff -urN wine-10.15/dlls/xinput1_3/Makefile.in b/dlls/xinput1_3/Makefile.in
--- wine-10.15/dlls/xinput1_3/Makefile.in	2026-03-11 20:17:01.134985603 +0800
+++ b/dlls/xinput1_3/Makefile.in	2026-03-11 21:22:09.126988859 +0800
@@ -1,6 +1,6 @@
 MODULE    = xinput1_3.dll
 IMPORTLIB = xinput
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32

 SOURCES = \
 	main.c \
diff -urN wine-10.15/dlls/xinput1_4/Makefile.in b/dlls/xinput1_4/Makefile.in
--- wine-10.15/dlls/xinput1_4/Makefile.in	2026-03-11 20:17:01.134985603 +0800
+++ b/dlls/xinput1_4/Makefile.in	2026-03-11 21:22:09.286988862 +0800
@@ -1,5 +1,5 @@
 MODULE    = xinput1_4.dll
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
 PARENTSRC = ../xinput1_3

 SOURCES = \
diff -urN wine-10.15/dlls/xinput9_1_0/Makefile.in b/dlls/xinput9_1_0/Makefile.in
--- wine-10.15/dlls/xinput9_1_0/Makefile.in	2026-03-11 20:17:01.134985603 +0800
+++ b/dlls/xinput9_1_0/Makefile.in	2026-03-11 21:22:09.442988864 +0800
@@ -1,4 +1,6 @@
 MODULE    = xinput9_1_0.dll
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
+PARENTSRC = ../xinput1_3

 SOURCES = \
 	main.c \
diff -urN wine-10.15/dlls/xinputuap/Makefile.in b/dlls/xinputuap/Makefile.in
--- wine-10.15/dlls/xinputuap/Makefile.in	2026-03-11 20:17:01.138985603 +0800
+++ b/dlls/xinputuap/Makefile.in	2026-03-11 21:22:09.810988871 +0800
@@ -1,5 +1,5 @@
 MODULE    = xinputuap.dll
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
 PARENTSRC = ../xinput1_3

 SOURCES = \