diff -urN wine-10.15/dlls/xinput1_3/Makefile.in b/dlls/xinput1_3/Makefile.in
--- wine-10.15/dlls/xinput1_3/Makefile.in	2026-03-11 20:17:01.134985603 +0800
+++ b/dlls/xinput1_3/Makefile.in	2026-03-11 21:22:09.126988859 +0800
@@ -1,6 +1,6 @@
 MODULE    = xinput1_3.dll
 IMPORTLIB = xinput
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32

 SOURCES = \
 	main.c \
diff -urN wine-10.15/dlls/xinput1_4/Makefile.in b/dlls/xinput1_4/Makefile.in
--- wine-10.15/dlls/xinput1_4/Makefile.in	2026-03-11 20:17:01.134985603 +0800
+++ b/dlls/xinput1_4/Makefile.in	2026-03-11 21:22:09.286988862 +0800
@@ -1,5 +1,5 @@
 MODULE    = xinput1_4.dll
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
 PARENTSRC = ../xinput1_3

 SOURCES = \
diff -urN wine-10.15/dlls/xinput9_1_0/Makefile.in b/dlls/xinput9_1_0/Makefile.in
--- wine-10.15/dlls/xinput9_1_0/Makefile.in	2026-03-11 20:17:01.134985603 +0800
+++ b/dlls/xinput9_1_0/Makefile.in	2026-03-11 21:22:09.442988864 +0800
@@ -1,4 +1,6 @@
 MODULE    = xinput9_1_0.dll
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
+PARENTSRC = ../xinput1_3

 SOURCES = \
 	main.c \
diff -urN wine-10.15/dlls/xinputuap/Makefile.in b/dlls/xinputuap/Makefile.in
--- wine-10.15/dlls/xinputuap/Makefile.in	2026-03-11 20:17:01.138985603 +0800
+++ b/dlls/xinputuap/Makefile.in	2026-03-11 21:22:09.810988871 +0800
@@ -1,5 +1,5 @@
 MODULE    = xinputuap.dll
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
 PARENTSRC = ../xinput1_3

 SOURCES = \
diff -urN wine-10.15/dlls/dinput/Makefile.in b/dlls/dinput/Makefile.in
--- wine-10.15/dlls/dinput/Makefile.in	2026-03-11 20:16:58.922985604 +0800
+++ b/dlls/dinput/Makefile.in	2026-03-11 21:21:17.762986366 +0800
@@ -1,6 +1,6 @@
 MODULE    = dinput.dll
 IMPORTLIB = dinput
-IMPORTS   = dinput dxguid uuid comctl32 ole32 user32 advapi32 hid setupapi
+IMPORTS   = dinput dxguid uuid comctl32 ole32 user32 advapi32 hid setupapi ws2_32
 EXTRADEFS = -DDIRECTINPUT_VERSION=0x0700

 SOURCES = \
@@ -15,4 +15,5 @@
 	joystick_hid.c \
 	keyboard.c \
 	mouse.c \
-	version.rc
+	version.rc \
+	gamepad.c
diff -urN wine-10.15/dlls/dinput8/Makefile.in b/dlls/dinput8/Makefile.in
--- wine-10.15/dlls/dinput8/Makefile.in	2026-03-11 20:16:58.934985604 +0800
+++ b/dlls/dinput8/Makefile.in	2026-03-11 21:21:52.742988434 +0800
@@ -1,6 +1,6 @@
 MODULE    = dinput8.dll
 IMPORTLIB = dinput8
-IMPORTS   = dinput8 dxguid uuid comctl32 ole32 user32 advapi32 hid setupapi
+IMPORTS   = dinput8 dxguid uuid comctl32 ole32 user32 advapi32 hid setupapi ws2_32
 EXTRADEFS = -DDIRECTINPUT_VERSION=0x0800
 PARENTSRC = ../dinput

@@ -16,4 +16,5 @@
 	joystick_hid.c \
 	keyboard.c \
 	mouse.c \
-	version.rc
+	version.rc \
+	gamepad.c
diff -urN wine-10.15/dlls/xinput1_1/Makefile.in b/dlls/xinput1_1/Makefile.in
--- wine-10.15/dlls/xinput1_1/Makefile.in	2026-03-11 20:17:01.130985603 +0800
+++ b/dlls/xinput1_1/Makefile.in	2026-03-11 21:22:08.878988854 +0800
@@ -1,5 +1,5 @@
 MODULE    = xinput1_1.dll
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
 PARENTSRC = ../xinput1_3

 SOURCES = \
diff -urN wine-10.15/dlls/xinput1_2/Makefile.in b/dlls/xinput1_2/Makefile.in
--- wine-10.15/dlls/xinput1_2/Makefile.in	2026-03-11 20:17:01.134985603 +0800
+++ b/dlls/xinput1_2/Makefile.in	2026-03-11 21:22:09.022988857 +0800
@@ -1,5 +1,5 @@
 MODULE    = xinput1_2.dll
-IMPORTS   = hid setupapi advapi32 user32
+IMPORTS   = hid setupapi advapi32 user32 ws2_32
 PARENTSRC = ../xinput1_3

 SOURCES = \